src/solvers/file_solution_history.C

```diff
@@ -32,29 +32,57 @@ FileSolutionHistory::~FileSolutionHistory ()
 }
 
 // This function finds, if it can, the entry where we're supposed to
-// be storing data
+// be storing data, leaves stored_sols unchanged if it cant find an entry
+// with the key corresponding to time.
 void FileSolutionHistory::find_stored_entry(Real time)
 {
   if (stored_solutions.begin() == stored_solutions.end())
     return;
 
   libmesh_assert (stored_sols != stored_solutions.end());
 
-  // We will use the map::lower_bound operation to find the least key, which
-  // is the least upper bound among all the keys for the time of our interest.
+  // We will use the map::lower_bound operation to find the key which
+  // is the least upper bound among all existing keys for time.
   // (key before map::lower_bound) < time < map::lower_bound, one of these
-  // will be within TOLERANCE of time (unless we are creating a new map entry)
+  // should be within TOLERANCE of time (unless we are creating a new map entry)
+  // If the lower bound iterator points to:
+  // begin -> we are looking for the solution at the initial time
+  // end -> we are creating a new entry
+  // anything else, we are looking for an existing entry
   stored_solutions_iterator lower_bound_it = stored_solutions.lower_bound(time);
 
-  // Set the stored sols iterator to whichever key is within TOLERANCE of time
+  // For the key right before the lower bound
+  stored_solutions_iterator lower_bound_it_decremented;
+
+  // If we are at end, we are creating a new entry, nothing more to do
```

Comment:
Is this correct?  If our last stored_solutions time is t, and we ask for t-epsilon^2, then the lower_bound is going to return end, even though we surely want t.

Reply:
In that case, we need to add an if check in SH::retrieve, which moves stored_sols to the entry before end, if the call find_stored_entry in retrieve ends up pointing stored_sols to the end.

Reply:
Could you explain?  Right now, if we find_stored_history(final_time-epsilon^2), it would fail to find it.  With my suggestion, it would find it.  Why would the latter result require an extra if check that the former would not?